### DISCUSSION TITLE
Managing the Helmholtz Equation of State

### DISCUSSION THREAD

**Discussion by Nicolas-Capmany** [CoolProp/CoolProp#2403] [Q&A]
Hello Dr Bell, 

I would like to ask you for your advice on writing a subroutine that uses the Helmholtz Equation of State. I am a PhD student looking into the thermodynamics of Hydrogen. As such, I've thoroughly reviewed my thermodynamics lectures: the criteria for Spontaneity, the Maxwell relations, etc...

As part of my PhD, I would like to write code for a subroutine that applies a multiparametric Equation of State, namely the *Helmholtz Equation of State for Parahydrogen, Normal Hydrogen and Orthohydrogen* proposed by *Leachman, Jacobsen, Lemmon and Penoncello*:

*Leachman, J. W., Jacobsen, R. T., Penoncello, S. G. and Lemmon, E. W. (2009) “Fundamental Equations of State for Parahydrogen, Normal Hydrogen, and Orthohydrogen.” doi: 10.1063/1.3160306.*

I would like to program this because I think it would be a good exercise to put all the thermodynamics concepts I just revised together and appreciate how the different terms are linked to each other. My main issue is that I don't know where to start, and I was wondering if there would be any structured approach you might suggest to follow or any references you believe might be helpful. I would like to learn how to implement multiparametric equations of state, and I couldn't think of any other person with better experience in this. 

I apologise for the inconvenience, I understand you have a busy schedule, and I appreciate your time on reading this. 

Kind regards,

**Comment by ibell** [comment]
What are you trying to achieve in the end? If you "just" want to do something like calculate phase equilibria, I can recommend you use the complex step derivative approach, described in this paper: https://pubs.acs.org/doi/abs/10.1021/acs.iecr.2c00237 . For more complex things, you are better off using existing libraries.

The EOS you are speaking of is already implemented in CoolProp, teqp, and REFPROP. But I agree with you that there is no better learning than hands-on and doing it yourself. And implementing the residual portion of the Helmholtz energy and doing other calculations with it is a great learning opportunity.

**Comment by Nicolas-Capmany** [comment]
Thank you very much for taking time to reply Dr Bell, 

What I would initially want to achieve is being able to relate the physical properties of hydrogen (i.e., Pressure, Density, Temperature, etc...). Basically being able to code a simple subroutine that would take in two intensive properties (e.g., pressure and density) and get out another (e.g., temperature). Just as if I were coding the ideal gas law to
 $$PV = nRT$$ 
into a subroutine to get any other variable from a combination of two intensive inputs. I was wondering how this could be done with a more complex EOS like Helmholtz. So essentially, yes, I would like to learn how to apply the residual part of the Helmholtz energy (since the remaining part is obtained from the ideal as law).  I understand this is the part I would need to learn to apply this equation of state to relate different fluid properties, including those for phase equilibrium and states inside the vapour dome. 

Just to be more clear, with respect to the physical properties, I wouldn't expect to code a routine that would get every possible combination of intensive properties to obtain a third. I would mainly be looking at obtaining enthalpies, entropies and vapour fractions ($\alpha$) from combinations of pressure, densities or temperatures for example. 

At a later stage (supposing I was succesful on the step to relate physical properties), I would try to get the transport propeties 
$(C_p, \mu, k_{cond})$ 
from the Helmholtz equation of state if possible too. 

I imagine it might be very naive from my side pretending I could use the Helmholtz EOS in a code of my own, but I thought I  should try to ask for indications. If I may address you somewhat informally, I think what you've done with CoolProp is **pretty cool** (no pun intended), and I would like just to get a hint of how it is to use multiparametric equations of state since I enjoy studying thermodynamics myself. Hopefully what I've written here will make some sense (I apologise if it doesn't). 

Again, thank you very much for your time. I appreciate you are quite busy yourself and I don't want to be a nuisance. 

Kind regards,

**Reply by ibell** [reply]
If you want to implement just the residual part, then that would cover $T$, $p$, $v$ inputs as options, as well as phase equilibria.

You cannot get (at least directly) transport properties from the EOS. But the specific heat is a thermodynamic property so that should be fine. I would recommend to read my new paper about teqp I referenced in the previous post. Also read the paper about CoolProp. And the one from REFPROP is very useful too: https://pubs.acs.org/doi/full/10.1021/acs.iecr.2c01427

**Comment by Nicolas-Capmany** [comment]
By the way I apologise if you saw this exact message published before from a user account named: **n1tr1le**, I was logged into this secondary account when I replied to the message instead of the original one I used to create this thread. I apologise for the confusion. 

Kind regards, 

Nicolas
